amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,548 @@
+        * ``nextpnr-gowin``
+        * ``gowin_pack``
+
+    The environment is populated by running the script specified in the environment variable
+    ``AMARANTH_ENV_APICULA``, if present.
+
+    Build products:
+        * ``{{name}}.fs``: binary bitstream.
+
+    .. rubric:: Gowin toolchain
+
+    Required tools:
+        * ``gw_sh``
+
+    The environment is populated by running the script specified in the environment variable
+    ``AMARANTH_ENV_GOWIN``, if present.
+
+    Build products:
+        * ``{{name}}.fs``: binary bitstream.
+
+    """
+
+    toolchain = None # selected when creating platform
+
+    device    = abstractproperty()
+    series    = abstractproperty()
+    subseries = abstractproperty()
+    voltage   = abstractproperty()
+    size      = abstractproperty()
+    package   = abstractproperty()
+    speed     = abstractproperty()
+    osc_div   = abstractproperty()
+
+    @property
+    def _part(self):
+            return "{}-{}{}{}{}".format(self.series, self.voltage,
+            self.size, self.package, self.speed)
+
+    osc_dev_mapping = {
+        "OSC": [
+            "GW1N-4", "GW1N-4B", "GW1N-4C", "GW1N-9", "GW1N-9C",
+            "GW1NR-4", "GW1NR-4B", "GW1NR-4C", "GW1NR-9", "GW1NR-9C",
+            "GW1NRF-4B",
+            "GW2A-18", "GW2A-18C", "GW2A-55", "GW2A-55C",
+            "GW2AN-55C",
+            "GW2AR-18", "GW2AR-18C",
+            "GW2ANR-18C"],
+        "OSCZ": [
+            "GW1NS-4", "GW1NS-4C",
+            "GW1NSR-4", "GW1NSR-4C",
+            "GW1NSER-4C",
+            "GW1NZ-1"],
+        "OSCF": [
+            "GW1NS-2", "GW1NS-2C",
+            "GW1NSE-2C",
+            "GW1NSR-2", "GW1NSR-2C"],
+        "OSCH": [
+            "GW1N-1", "GW1N-1S",
+            "GW1NR-1"],
+        "OSCO": [
+            "GW1N-2", "GW1N-1P5", "GW1N-2B", "GW1N-1P5B",
+            "GW1NR-2", "GW1NR-2B"],
+        "OSCW": [
+            "GW2AN-18X", "GW2AN-9X"],
+    }
+
+    # Each on-chip oscillator has unique properties
+    @property
+    def _osc_props(self):
+        props = {
+            "freq": 0,
+            "div_min": 2,
+            "div_max": 128,
+            "div_step": 2
+        }
+        osc = self._oscillator
+        if osc == "OSC":
+            if "-4" in self.device:
+                props["freq"] = 210000000
+            else:
+                props["freq"] = 250000000
+        elif osc in ["OSCZ", "OSCO"]:
+            props["freq"] = 250000000
+        elif osc in ["OSCF", "OSCH"]:
+            props["freq"] = 240000000
+        elif osc == "OSCW":
+            props["freq"] = 200000000
+        else:
+            assert False
+        return props
+
+
+    @property
+    def _oscillator(self):
+        oscillators = []
+        for osc,devs in self.osc_dev_mapping.items():
+            if self.device in devs:
+                oscillators.append(osc)
+        # should this be unique?
+        if len(oscillators) != 1:
+            assert False
+        return oscillators[0]
```

Comment:
I think there's no especially good reason to do this dict inversion every time at runtime. Can you put the inverted dict in the sources?

Reply:
Yes. Done.